drivers/note/notesnap_driver.c

```diff
@@ -0,0 +1,422 @@
+{
+  notesnap_common(tcb, locked ? NOTE_PREEMPT_LOCK : NOTE_PREEMPT_UNLOCK, 0);
+}
+#endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_CSECTION
+static void notesnap_csection(FAR struct note_driver_s *drv,
+                              FAR struct tcb_s *tcb, bool enter)
+{
+  notesnap_common(tcb, enter ? NOTE_CSECTION_ENTER : NOTE_CSECTION_LEAVE, 0);
+}
+#endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SPINLOCKS
+static void notesnap_spinlock(FAR struct note_driver_s *drv,
+                              FAR struct tcb_s *tcb,
+                              FAR volatile void *spinlock, int type)
+{
+  notesnap_common(tcb, type, (uintptr_t)spinlock);
+}
+#endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
+static void notesnap_irqhandler(FAR struct note_driver_s *drv, int irq,
+                                FAR void *handler, bool enter)
+{
+  notesnap_common(this_task(), enter ? NOTE_IRQ_ENTER : NOTE_IRQ_LEAVE, irq);
+}
+#endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+static void notesnap_syscall_enter(FAR struct note_driver_s *drv, int nr,
+                                   int argc, va_list *ap)
+{
+  notesnap_common(this_task(), NOTE_SYSCALL_ENTER, nr);
+}
+
+static void notesnap_syscall_leave(FAR struct note_driver_s *drv, int nr,
+                                   uintptr_t result)
+{
+  notesnap_common(this_task(), NOTE_SYSCALL_LEAVE, nr);
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: notesnap_register
+ ****************************************************************************/
+
+int notesnap_register(void)
+{
+  g_notesnap.dumping = false;
+  return note_driver_register(&g_notesnap.driver);
+}
+
+/****************************************************************************
+ * Name: notesnap_dump
+ ****************************************************************************/
+
+void notesnap_dump(FAR struct lib_outstream_s *stream)
+{
+  size_t i;
+  size_t index = g_notesnap.index % CONFIG_DRIVERS_NOTESNAP_NBUFFERS;
+  FAR struct notesnap_chunk_s *note;
+
+#ifdef CONFIG_DRIVERS_NOTESNAP_PERFCOUNT
+  uint32_t elapsed;
+  uint32_t lasttick = g_notesnap.buffer[index].tick;
+  struct timespec time;
+  struct timespec current_time;
+  struct timespec last_time =
+  {
+    0
+  };
+#endif
+
+  /* Stop recording while dumping */
+
+  g_notesnap.dumping = true;
+
+  for (i = index; i != index - 1;
+       i == CONFIG_DRIVERS_NOTESNAP_NBUFFERS - 1 ? i = 0 : i++)
+    {
+      note = &g_notesnap.buffer[i];
+
+#ifdef CONFIG_DRIVERS_NOTESNAP_PERFCOUNT
+      elapsed = note->tick < lasttick ? note->tick + UINT32_MAX - lasttick
+                                      : note->tick - lasttick;
+      up_perf_convert(elapsed, &time);
+      clock_timespec_add(&last_time, &time, &current_time);
```

Comment:
change:
clock_timespec_add(&last_time, &time, &last_time);
and remove current_time